src/PeerSelectState.h

```diff
@@ -114,24 +191,45 @@ class PeerSelector: public Dns::IpReceiver
     void selectAllParents();
     void selectPinned();
 
-    void addSelection(CachePeer*, const hier_code);
-
-    void resolveSelected();
+    void sendNextPeer();
 
     static IRCB HandlePingReply;
     static ACLCB CheckAlwaysDirectDone;
     static ACLCB CheckNeverDirectDone;
     static EVH HandlePingTimeout;
 
 private:
+    /// Selection states
+    enum {
+        DoPreselection, ///< check for pinned, peers digest and closed parents
+        DoPing, ///< ping peers
+        DoFinal, ///< all of the remaining valid peers
+        DoFinished ///< all steps finished
+    } selectionState = DoPreselection;
+
     allow_t always_direct;
     allow_t never_direct;
     int direct;   // TODO: fold always_direct/never_direct/prefer_direct into this now that ACL can do a multi-state result.
-    size_t foundPaths = 0; ///< number of unique destinations identified so far
-    ErrorState *lastError;
 
+    int foundPeers = 0;
     FwdServer *servers; ///< a linked list of (unresolved) selected peers
 
+    /// The member of servers list which is currently processed and checked
+    /// before sent to caller
+    FwdServer *currentServer = nullptr;
+
+    AsyncCall::Pointer callback_; ///< caller callback
+
+    /// Temporary list with candidate peers to ping. They will be ACL checked
+    /// before added to peersToPing list
+    std::vector<CbcPointer<CachePeer> > candidatePingPeers;
+
+    /// The list with available peers for pinging
+    std::vector<CbcPointer<CachePeer> > peersToPing;
+
+    /// Cache for peers ACL checks
+    std::vector<std::pair<CbcPointer<CachePeer>, allow_t> > aclPeersCache;
```

Comment:
I have dilemma here. I selected std::vector because (1) I am expecting many PeerSelectState classes in squid memory, (2) std::vector probably is uses less memory than std::map and (3) I am expecting only few ACL checks results in this cache. But I need suggestion here.

Reply:
I suggest the following steps:

1. Document your motivation in a source comment. Just enumerate the pros you gave in the above GitHub comment. You can also add cons like linear searches and expensive erasures.
2. Continue to use std::vector as is for now.
3. Once all other problems are sorted out, we will come back to this and decide whether std::vector is the best structure for this data member. Add a TODO comment to remind us. For example, `TODO: Decide whether std::vector is the best storage type for this cache`.

Reply:
> ... add cons like linear searches and expensive erasures
Only linear search, we are never deleting.